src/fmureceiver.cpp

```diff
@@ -0,0 +1,307 @@
+        {
+            fmi2_boolean_t newStep = fmi2_true;
+            time_t timeC = time(NULL);
+            fmiStatus_ = fmi2_import_do_step(fmu_, (fmi2_real_t)timeC, hStep_, newStep);
+
+            osi::SensorData sd;
+            if (get_fmi_sensor_data_in(sd))
+            {
+                qDebug() << "FMU SensorData Receive data";
+                msgReceived = true;
+                emit MessageReceived(sd, currentDataType_);
+            }
+            else
+            {
+//                qDebug() << "FMU SensorData receiving error";
+            }
+        }
+
+        if (!msgReceived)
+        {
+            QThread::msleep(5);
+        }
+    }
+
+    isThreadTerminated_ = true;
+}
+
+bool
+FMUReceiver::initializeFMUWrapper()
+{
+    // TODO: config or use default values?
+    callbacks_.malloc = malloc;
+    callbacks_.calloc = calloc;
+    callbacks_.realloc = realloc;
+    callbacks_.free = free;
+    callbacks_.logger = fmuWrapperLogger;
+    if (logLevel_ == LogLevel::Warn)
+    {
+        callbacks_.log_level = jm_log_level_warning;
+    }
+    else
+    {
+        callbacks_.log_level = jm_log_level_debug;
+    }
+    callbacks_.context = 0;
+
+    context_ = fmi_import_allocate_context(&callbacks_);
+    fmi_version_enu_t version = fmi_import_get_fmi_version(context_, FMUPath_.c_str(), tmpPath_.c_str());
+
+    if (version != fmi_version_2_0_enu)
+    {
+        qDebug() << "The code only supports version 2.0";
+        return false;
+    }
+    return true;
+}
+
+bool
+FMUReceiver::importFMU()
+{
+    // TODO: config or use default values?
+    callBackFunctions_.logger = fmi2_log_forwarding;
+    callBackFunctions_.allocateMemory = calloc;
+    callBackFunctions_.freeMemory = free;
+    callBackFunctions_.componentEnvironment = fmu_;
+
+    jmStatus_ = fmi2_import_create_dllfmu(fmu_, fmi2_fmu_kind_cs, &callBackFunctions_);
+
+    if (jmStatus_ == jm_status_error)
+    {
+        return false;
+    }
+    return true;
+}
+
+bool
+FMUReceiver::initializeFMU()
+{
+    // TODO: config or use default values?
+    // TODO: alternative solution for start values & verify parameters for setup and initiliazation functions
+    // start values
+    fmi2_string_t instanceName = "Test CS model instance";
+
+    fmi2_string_t fmuLocation = "";
+    fmi2_boolean_t visible = fmi2_false;
+    fmi2_real_t relativeTol = 1e-4;
+
+    tStart_ = 0;
+    tCurrent_ = tStart_;
+    fmi2_boolean_t StopTimeDefined = fmi2_false;
+
+    // Do we need it?
+    fmi2_string_t fmuGUID;
+    fmuGUID = fmi2_import_get_GUID(fmu_);
+
+    jmStatus_ = fmi2_import_instantiate(fmu_, instanceName, fmi2_cosimulation, fmuLocation, visible);
+    fmiStatus_ = fmi2_import_setup_experiment(fmu_, fmi2_true, relativeTol, tStart_, StopTimeDefined, tEnd_);
+    fmiStatus_ = fmi2_import_enter_initialization_mode(fmu_);
+    fmiStatus_ = fmi2_import_exit_initialization_mode(fmu_);
+
+    vr_[FMI_INTEGER_SENSORDATA_IN_BASELO_IDX] = FMI_INTEGER_SENSORDATA_IN_BASELO_IDX;
```

Comment:
The variable accesses should be done via the specified variable names, not through hard-coded VRs (these can change at any time, only the Variable names are guaranteed to remain constant through the OSMP spec). E.g. variables can be looked up via fmi2_import_get_variable_by_name and their VR via fmi2_import_get_variable_vr, all to be done at load-time of the FMU.